src/core/ext/xds/xds_client_grpc.cc

```diff
@@ -230,14 +236,28 @@ void SetXdsFallbackBootstrapConfig(const char* config) {
 }  // namespace grpc_core
 
 // The returned bytes may contain NULL(0), so we can't use c-string.
-grpc_slice grpc_dump_xds_configs(void) {
+grpc_slice grpc_dump_xds_configs(void) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
   grpc_core::ExecCtx exec_ctx;
   auto xds_client = grpc_core::GrpcXdsClient::GetOrCreate(
       grpc_core::ChannelArgs(), "grpc_dump_xds_configs()");
-  if (!xds_client.ok()) {
-    // If we aren't using xDS, just return an empty string.
-    return grpc_empty_slice();
+  upb::Arena arena;
+  // Following two containers should survive till serialization
+  std::vector<std::unique_ptr<std::string>> string_pool;
+  auto response = envoy_service_status_v3_ClientStatusResponse_new(arena.ptr());
+  // If we aren't using xDS, just return a response with empty config. This is
+  // consistent with older gRPC implementation
+  auto client_config = envoy_service_status_v3_ClientStatusResponse_add_config(
+      response, arena.ptr());
+  std::vector<std::unique_ptr<grpc_core::MutexLock>> client_locks;
```

Comment:
Interesting!  I thought that `reserve()` just allocated space without instantiating the objects, but it looks like it does actually need to default-construct them.

Note that `emplace_back()` does not require a move constructor.  The whole point of `emplace_back()` over `push_back()` is that it instantiates the object in place rather than moving or copying.  So the only problem here is actually the `reserve()`.

Thinking about this further, I think that once we actually have multiple `XdsClient` instances, we're going to have to construct a vector of refs to those instances anyway, and once we do that, it doesn't really make sense to also construct a vector of locks.  I think this is one of those rare cases where it will make sense to manually lock and unlock instead of using `MutexLock`:

```
// Grab a ref to every XdsClient instance.
std::vector<RefCountedPtr<GrpcXdsClient>> xds_clients;
{
  MutexLock lock(g_mu);
  xds_clients.reserve(g_xds_client_map->size());
  for (const auto& p : *g_xds_client_map) {
    const std::string& client_scope = p.first;
    auto xds_client = p.second->RefIfNonZero(DEBUG_LOCATION, "csds");
    if (xds_client != nullptr) {
      xds_clients.push_back(xds_client.TakeAsSubclass<GrpcXdsClient>());
    }
  }
}
// Construct response proto.
upb::Arena arena;
auto* response = envoy_service_status_v3_ClientStatusResponse_new(arena.ptr());
// For each XdsClient, acquire the lock and append its config to the response proto.
// Note that we continue to hold the lock until after we serialize,
// so that we don't need to copy data.
std::set<std::string> string_pool;
for (auto& xds_client : xds_clients) {
  auto* client_config = envoy_service_status_v3_ClientStatusResponse_add_config(
      response, arena.ptr());
  xds_client->mu()->Lock();
  xds_client->DumpClientConfig(&string_pool, arena.ptr(), client_config);
}
// Serialize the upb message to bytes.
size_t output_length;
char* output = envoy_service_status_v3_ClientStatusResponse_serialize(
    response, arena.ptr(), &output_length);
grpc_slice result = grpc_slice_from_cpp_string(std::string(output, output_length));
// Release mutexes.
for (auto& xds_client : xds_clients) {
  xds_client->mu()->Unlock();
}
return result;
```

We can use basically this same pattern now, even though there's only one `XdsClient`.